plat/common/include/uk/plat/common/memory.h

```diff
@@ -112,6 +117,75 @@ ukplat_memregion_list_insert(struct ukplat_memregion_list *list,
 	return (int)i;
 }
 
+/**
+ * Insert a new region into memory region list. This extends
+ * ukplat_memregion_list_insert to carve out the area of any pre-existing
+ * overlapping regions. The virtual addresses will leave out any carved out
+ * region.
+ * @param list
+ *   The memory region list to insert the range into
+ * @param mrd
+ *   The memory range to insert
+ * @param min_size
+ *   The minimum size a inserted region has to have. Regions smaller than this
+ *   size won't be added. Settings this parameter zero will disable this
+ *   behavior.
+ * @returns
+ *   zero if the operation was successful, an negative errno otherwise
+ */
+static inline int
+ukplat_memregion_list_insert_split_phys(struct ukplat_memregion_list *list,
+					const struct ukplat_memregion_desc *mrd,
+					__sz min_size)
+{
+	struct ukplat_memregion_desc *mrdp;
+	struct ukplat_memregion_desc mrdc;
+	__paddr_t pstart, pend;
+	__vaddr_t voffset;
+	int i;
+	int rc;
+
+	voffset = mrd->vbase - mrd->pbase;
+	pstart = mrd->pbase;
+	pend = mrd->pbase + mrd->len;
+
+	memcpy(&mrdc, mrd, sizeof(mrdc));
+
+	for (i = 0; i < (int)list->count; i++) {
+		ukplat_memregion_get(i, &mrdp);
+		if (!ukplat_memregion_desc_overlap(mrdp, pstart, pend))
+			continue;
+
+		if (!mrdp->type)
```

Comment:
By the way you seem to be checking here for a non-zero mrd type. But what if the region you are about to split is of a type that would require a higher priority than the one you are comparing it to? For example, what if you are adding a mrd of type RESERVED whose pstart < mrdp->pbase and whose pend > mrdp->pbase + mrdp->len. Then you would have to remove mrdp entirely from the list, since the reserved mrd you are about to add takes priority.